### ISSUE TITLE
Automatically service registration and discovery support in cloud enviroment

### ISSUE THREAD

**Issue by jonefeewang** [line/armeria#331]
when use armeria server in cloud environment, like mesos . we should provide a easy way to let every server node automatically regist itself  in a  distribute coordinate service like ZooKeeper. Currently, i am planning to implement a ZooKeeper `ServerListener`, so that users could easily add this listener to achieve a service registration.

if the client is also armeria client, we could provide unique service alias to make the service discovery extremely easily. for example, when the server is deployed in a cloud ,it automatically register itself under a ZooKeeper zNode(we needn't let the user set the node name ,just chose a random name for it.) and identified this service as a alias like `com.abc.mobiledep.foo_http_service`, the client could simply specify " i want to use `com.abc.mobledep.foo_http_servie` " to connect to the server.

By this way , when the user deploy the server, he just need to add a server listener to specify ZooKeeper service url etc., and specify a unique "service name(com.abc.mobiledep.foo_http_service)", the client just need to create a `ZooKeeperEndGroup`, specify  the corresponding ZooKeeper connection info and specify the unique "service name(com.abc.mobiledep.foo_http_service)"

**Comment by trustin** [comment]
Adding a `ServerListener` implementation sounds like a good idea.

To map a Service into a zNode path, how about defining the following interface?

```java
public interface ZooKeeperNodeMapper extends Function<ServiceConfig, Optional<String>> {
    ZooKeeperNodeMapper DEFAULT = ...;

    @Override
    public Optional<String> apply(ServiceConfig cfg); // Returns Optional.empty() for the excluded services
}
```

When creating the `ServerListener` implementation, a user could specify what ZooKeeperNodeMapper to use. What do you think?

**Comment by jonefeewang** [comment]
@trustin sorry for this late reply, it is a good idea, i will make a draft implementation first.

**Comment by trustin** [comment]
@jonefeewang Gentle ping. How's it going? :-)

**Comment by jonefeewang** [comment]
Sorry for this delayed feature, I am currently trapped by company project.

Hope complete this feature in two weeks.

Thanks for your patients!


From:  Trustin Lee <notifications@github.com>
Reply-To:  line/armeria
<reply+002e461760d3e3493d811d2a5fe44502b72b2b5b75bab4eb92cf0000000114672f959
2a169ce0b431441@reply.github.com>
Date:  Tuesday, December 13, 2016 at 11:40
To:  line/armeria <armeria@noreply.github.com>
Cc:  王 军飞 <jonefeewang@gmail.com>, Mention <mention@noreply.github.com>
Subject:  Re: [line/armeria] Automatically service registration and
discovery support in cloud enviroment (#331)

**Comment by trustin** [comment]
> Thanks for your patience!

No worries. Please take your time. :-)

**Comment by trustin** [comment]
Fixed via #357
